vignettes/fims-path-maturity.Rmd

```diff
@@ -0,0 +1,614 @@
+
+There are some exceptions where `{TMB}` specific code is referenced within the `fims` namespace. In these cases, code is written within an `#ifdef TMB_MODEL` wrapper, which means the code is only defined if `{TMB}` is being used. Given the addition of a new platform, eg. stan, a new wrapper could be added to define platform specific code for these sections. 
+
+For example, the definition of data types in [interface.hpp](https://github.com/NOAA-FIMS/FIMS/blob/main/inst/include/interface/interface.hpp) under [inst/include/interface](https://github.com/NOAA-FIMS/FIMS/tree/main/inst/include/interface) will always be platform dependent as each platform has specific requirements for how the data types are defined. Whenever a new platform is added to `{FIMS}`, we will need to set up the platform specific data type definitions:
+```{Rcpp,  class.source = "fimschunk", eval = FALSE}
+namespace fims {
+
+#ifdef TMB_MODEL
+/**
+ *  @brief ModelTraits class that contains the DataVector
+ * and ParameterVector types.
+ */
+template <typename T>
+struct ModelTraits {
+  typedef typename CppAD::vector<T> DataVector;      /**< This is a vector
+        of the data that is differentiable */
+  typedef typename CppAD::vector<T> ParameterVector; /**< This is a
+  vector of the parameters that is differentiable */
+  typedef typename tmbutils::vector<T>
+      EigenVector; /**< This is a vector as defined in TMB's namespace Eigen */
+};
+#endif /* TMB_MODEL */
+}  // namespace fims
+
+//not developed yet
+#ifdef STAN_MODEL
+//stan specific definitions go here
+#endif
+```
+
+### Getting back to the maturity example...
+Each Rcpp interface object includes an `add_to_fims_tmb()` function. There are two shared pointers set up within this function, one to link each Rcpp interface object (e.g. LogisticMaturityInterface) to the `Information` class in the `fims` namespace defined in [information.hpp](https://github.com/NOAA-FIMS/FIMS/blob/main/inst/include/common/information.hpp) under [inst/include/common](https://github.com/NOAA-FIMS/FIMS/tree/main/inst/include/common) and one to link to the matching class in the `fims` namespace. In our maturity example, this would be the `LogisticMaturity` class in the `fims` namespace defined in [logistic.hpp](https://github.com/NOAA-FIMS/FIMS/blob/main/inst/include/population_dynamics/maturity/functors/logistic.hpp) under [inst/include/population_dynamics/maturity/functors](https://github.com/NOAA-FIMS/FIMS/tree/main/inst/include/population_dynamics/maturity/functors).
+
+
+Within the rcpp_maturity.hpp file, we can link to the `fims::Information` class in order to register maturity parameters and specify whether or not they are random effects:
+
+```{Rcpp, class.source = "rcppchunk", eval = FALSE}
+//file: rcpp_maturity.hpp
+
+//d0 is a shared pointer that points to fims::Information
```

Comment:
suggestion: `d0` is not described in the text but it is used 8 times in the code within the is vignette. I think readers would benefit from a bit of a road map regarding what d0 is and how you chose to use the name lm0 and how the latter name is specific to logistic maturity and if you were to implement a new module you would name that something different, e.g., sl0 for logistic selectivity.

Reply:
I expanded the paragraph above this code chunk to read:

Within the rcpp_maturity.hpp file, we can link to the `fims::Information` class in order to register maturity parameters and specify whether or not they are random effects. We do this by setting up two pointers in the interface: d0 which points to information, and lm0, which points to the logistic maturity module. 

I am reluctant to go into details on the naming convention as it doesn't always make sense will likely change soon.